packages/jsdoc-plugins/lib/relation-fixer/buildrelations.js

```diff
@@ -2,132 +2,163 @@
  *
- * @param {Array.<Doclet>} originalDoclets
+ * @param {Array.<Doclet>} doclets
  * @returns {Array.<Doclet>}
  */
-function addRelationArrays( originalDoclets ) {
-	const clonedDoclets = cloneDeep( originalDoclets );
+function buildRelations( doclets ) {
+	// Preserve original doclets from modification for easier testing.
+	doclets = cloneDeep( doclets );
+
+	/**
+	 * Doclets grouped by their longnames.
+	 */
 	const docletCollection = new DocletCollection();
 
-	for ( const doclet of clonedDoclets ) {
-		// Group doclets by longname.
+	for ( const doclet of doclets ) {
 		docletCollection.add( doclet.longname, doclet );
 	}
 
-	// Doclets for which we want to create relation arrays.
-	// We want classes, interfaces and mixins.
-	const subjectDoclets = docletCollection.getAll().filter( item => {
-		return item.kind === 'class' || item.kind === 'interface' || item.kind === 'mixin' || item.kind === 'typedef';
+	/**
+	 * An array of doclets, for which we want to create relation arrays.
+	 */
+	const subjectDoclets = doclets.filter( item => {
+		return (
+			item.kind === 'class' ||
+			item.kind === 'interface' ||
+			item.kind === 'mixin' ||
+			item.kind === 'typedef'
+		);
 	} );
 
-	subjectDoclets.forEach( d => {
-		const related = getAncestors( docletCollection, d, {
+	for ( const doclet of subjectDoclets ) {
+		const related = getAncestors( docletCollection, doclet, {
 			relations: [ 'augments', 'implements', 'mixes' ]
 		} );
 
 		// Remove duplicates.
 		for ( const relation of Object.keys( related ) ) {
-			related[ relation ] = Array.from( new Set( related[ relation ] ) );
+			related[ relation ] = uniq( related[ relation ] );
 		}
 
-		Object.assign( d, related );
-	} );
-
-	subjectDoclets.forEach( d => {
-		const descendants = Array.from( new Set( getDescendants( subjectDoclets, d ) ) );
+		Object.assign( doclet, related );
+	}
 
-		Object.assign( d, { descendants } );
-	} );
+	for ( const doclet of subjectDoclets ) {
+		doclet.descendants = uniq( getDescendants( subjectDoclets, doclet ) );
+	}
 
-	return docletCollection.getAll();
+	return doclets;
 }
 
-// Gets longnames of currentDoclet's ancestors (classes it extends, interfaces it implements and so on).
-//
-// @param {DocletCollection} docletCollection
-// @param {Doclet} currentDoclet
-// @param {Array} options.relations Array of relation names which should be used
-// @returns {Object} An object containing arrays of ancestors' longnames
+/**
+ * Gets long names of the current doclet ancestors (classes it extends, interfaces it implements and so on).
+ *
+ * @param {DocletCollection} docletCollection Doclets grouped by their longnames
+ * @param {Doclet} currentDoclet
+ * @param {Object} options
+ * @param {Array.<'augments'|'implements'|'mixes'>} options.relations An array of relation names which should be used.
+ * @returns {{ [relation: string ]: Array.<String>}} An object containing arrays of ancestors' longnames.
```

Comment:
Just `@returns {Object} ...`

Reply:
It's more precise

Reply:
"more precise" !== proper definition in terms of our code style. We should use `Object` or create another custom type.